crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2145,6 +2138,83 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Return the outcome of calling an class/instance attribute of this type
+    /// using descriptor protocol.
+    ///
+    /// `receiver_ty` must be `Type::Instance(_)` or `Type::ClassLiteral`.
+    ///
+    /// TODO: handle `super()` objects properly
+    #[must_use]
+    fn call_bound(
+        self,
+        db: &'db dyn Db,
+        receiver_ty: &Type<'db>,
+        arguments: &CallArguments<'_, 'db>,
+    ) -> CallOutcome<'db> {
+        debug_assert!(receiver_ty.is_instance() || receiver_ty.is_class_literal());
+
+        match self {
+            Type::FunctionLiteral(..) => {
+                // Functions are always descriptors, so this would effectively call
+                // the function with the instance as the first argument
+                self.call(db, &arguments.with_self(*receiver_ty))
+            }
+
+            object_ty @ (Type::Instance(_) | Type::ClassLiteral(_)) => {
+                match object_ty.member(db, "__get__") {
+                    Symbol::Type(get_method, _) => {
+                        let descr_result = match receiver_ty {
+                            Type::Instance(_) => {
+                                let receiver_class = receiver_ty.to_meta_type(db);
+
+                                get_method.call(
+                                    db,
+                                    &CallArguments::positional([*receiver_ty, receiver_class]),
+                                )
+                            }
+                            Type::ClassLiteral(_) => get_method.call(
+                                db,
+                                &CallArguments::positional([
+                                    KnownClass::NoneType.to_instance(db),
+                                    object_ty,
+                                ]),
+                            ),
+                            _ => unreachable!(),
+                        };
+
+                        match descr_result {
+                            CallOutcome::Callable { binding } => {
+                                binding.return_type().call(db, arguments)
+                            }
+                            _ => descr_result,
+                        }
+                    }
+                    Symbol::Unbound => {
+                        // Not a descriptor, so call without passing the instance
+                        self.call(db, arguments)
+                    }
+                }
```

Comment:
Since we don't have tests for descriptor protocol (and this PR doesn't add any), and we may want to design our descriptor protocol support a bit more holistically (not just specific to bound calls), I think we should not add support for it here. Instead we should just simplify this code to assume non-descriptor, with a TODO comment for descriptor support.
```suggestion
                // TODO descriptor protocol. For now, assume non-descriptor and call without `self` argument.
                self.call(db, arguments)
```

Reply:
I'd love to work on the descriptor protocol as a whole, would be fun and will allow me to better review more of the red_knot codebase. If you are fine with that, I can open a separate draft PR and outline a plan.

One obvious place to start would be: https://github.com/astral-sh/ruff/blob/0529ad67d76d01c5ca586ad9b3db8b37c1a930dc/crates/red_knot_python_semantic/src/types.rs#L4211-L4219  and implement proper handling of `property` instead of a current "workaround". This would serve as a test. + I can make a special mutest suite for descriptors in general.

As for my code here, I do not think that there is harm in leaving it, as it will work for cases when `__get__` method is defined explicitly already, but since you suggest removing it, I'll just stash it and see if it eventually survives.

Reply:
Thanks for removing this for now. I agree it's possible that later it may come back in the same or similar form, but I would rather it arrive along with tests exercising it.

I think @sharkdp was planning to tackle descriptors soon, but may be open to you working on that instead, freeing him up to tackle something else? Let's wait for him to respond before you start on it, in case he has already started on it.

It would definitely be good to open an issue with a proposed design first, so we can discuss, before opening a PR; this may save some implementation work.

Reply:
@carljm @sharkdp I can create an outline with items that needs to be tackled in principle, a proposed step-by-step plan and draft design (or at least some thinking around it). Shall I use discussions or an issue?

Reply:
That sounds like a great approach! We usually use issues, not discussions. I think @sharkdp was planning to create a set of issues tomorrow for various follow-ups to his recent work on instance attributes, including descriptor protocol as one of those.

Reply:
> @carljm @sharkdp I can create an outline with items that needs to be tackled in principle, a proposed step-by-step plan and draft design (or at least some thinking around it). Shall I use discussions or an issue?

@mishamsk If you want to work on descriptors, we would definitely appreciate it! Starting with a separate design step sounds like a good idea. Please feel free to add comments to this issue: https://github.com/astral-sh/ruff/issues/15966

I will also set up an initial draft for a descriptor test suite that we can detail further while we work on the design and the implementation.